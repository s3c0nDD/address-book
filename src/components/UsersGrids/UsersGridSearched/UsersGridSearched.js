import React, { memo, useCallback, useEffect } from 'react';
import PropTypes from 'prop-types';
import { makeStyles } from '@material-ui/core/styles';

import Box from '@material-ui/core/Box';

import UsersGridLayout from '../UsersGridLayout/UsersGridLayout';
import UsersGridLayoutMessage from '../UsersGridLayout/UsersGridLayoutMessage';

const useStyles = makeStyles(theme => ({
  root: {
    padding: `0 ${theme.spacing(4)}px`
  },
  messageEmpty: {
    marginTop: theme.spacing(4)
  }
}));

const shouldNotRerender = (prevProps, nextProps) => {
  return prevProps.users === nextProps.users
    && prevProps.show === nextProps.show
    && prevProps.processing === nextProps.processing;
};

const UsersGridSearched = memo(({ processing, show, users, onOpenModal, ...props }) => {
  const classes = useStyles();

  const scrollToTop = useCallback(() => {
    window.scroll(0,0);
  }, []);

  useEffect(() => {
    scrollToTop();
  }, [users, scrollToTop]);

  const shouldShowNoResultMessage = (!users?.length && !processing);

  return (
    <>
      {show && (
        <Box
          className={classes.root}
          {...props}
        >
          <UsersGridLayout
            users={users}
            onOpenModal={onOpenModal}
          />
          {shouldShowNoResultMessage && (
            <UsersGridLayoutMessage
              className={classes.messageEmpty}
              message={'No matching results'}
            />
          )}
        </Box>
      )}
    </>
  );
}, shouldNotRerender);

export default UsersGridSearched;

UsersGridSearched.propTypes = {
  processing: PropTypes.bool.isRequired,
  show: PropTypes.bool,
  users: PropTypes.array.isRequired,
  onOpenModal: PropTypes.func.isRequired,
  props: PropTypes.arrayOf(PropTypes.any)
};