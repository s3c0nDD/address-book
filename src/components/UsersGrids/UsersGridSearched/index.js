import UsersGridSearched from './UsersGridSearched';

export default UsersGridSearched;