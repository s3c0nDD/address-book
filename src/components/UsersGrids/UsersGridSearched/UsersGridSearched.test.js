import React from 'react';
import { createMount, createShallow } from '@material-ui/core/test-utils';

import userMockData from '../../mocks/userMock';
import UsersGridSearched from './UsersGridSearched'

describe('UsersGridSearched', () => {
  let shallow;
  let mount;
  let wrapper;
  let component;
  let openModalMock;
  let scrollMock;
  let usersMock = [{ ...userMockData }, { ...userMockData, id: { value: 'foobaz' }}];

  beforeAll(() => {
    shallow = createShallow();
    mount = createMount();
    scrollMock = jest.fn();
    jest.spyOn(window, 'scroll').mockImplementation(scrollMock);
  });

  afterAll(() => {
    window.scroll.mockClear();
  });

  beforeEach(() => {
    openModalMock = jest.fn();
    component = (
      <UsersGridSearched
        show={true}
        processing={false}
        users={usersMock}
        onOpenModal={openModalMock}
      />
    );
    wrapper = shallow(component);
  });

  it('shallowly renders and matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render null when prop show is not truthy', () => {
    wrapper = mount(
      <UsersGridSearched
        show={false}
        processing={false}
        users={usersMock}
        onOpenModal={openModalMock}
      />
    );
    expect(wrapper.html()).toBeNull();
  });
});