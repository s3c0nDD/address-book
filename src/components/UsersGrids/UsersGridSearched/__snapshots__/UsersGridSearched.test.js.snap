// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersGridSearched shallowly renders and matches snapshot 1`] = `
<Fragment>
  <Styled(MuiBox)
    className="makeStyles-root-1"
  >
    <Memo()
      onOpenModal={[MockFunction]}
      users={
        Array [
          Object {
            "cell": "111-222-333",
            "email": "johndoe123@google.com",
            "id": Object {
              "value": "foobar",
            },
            "location": Object {
              "city": "Warsaw",
              "country": "Poland",
              "postcode": 56789,
              "state": "Masovian",
              "street": Object {
                "name": "Big Street",
                "number": 98,
              },
            },
            "login": Object {
              "username": "johndoe123",
            },
            "name": Object {
              "first": "John",
              "last": "Doe",
              "title": "Mr",
            },
            "phone": "000-111-222",
            "picture": Object {
              "large": "https://wwww.image.com/data/1",
            },
          },
          Object {
            "cell": "111-222-333",
            "email": "johndoe123@google.com",
            "id": Object {
              "value": "foobaz",
            },
            "location": Object {
              "city": "Warsaw",
              "country": "Poland",
              "postcode": 56789,
              "state": "Masovian",
              "street": Object {
                "name": "Big Street",
                "number": 98,
              },
            },
            "login": Object {
              "username": "johndoe123",
            },
            "name": Object {
              "first": "John",
              "last": "Doe",
              "title": "Mr",
            },
            "phone": "000-111-222",
            "picture": Object {
              "large": "https://wwww.image.com/data/1",
            },
          },
        ]
      }
    />
  </Styled(MuiBox)>
</Fragment>
`;
